src/lsm/scan_tree.zig

```diff
@@ -271,6 +288,73 @@ pub fn ScanTreeType(
                 },
                 .needs_data => return error.ReadAgain,
                 .buffering => unreachable,
+                .aborted => return null,
+            }
+        }
+
+        /// Probing a scan modifies the key_min/key_max range, either moving the scan if the next
+        /// key is already buffered, or causing it to read again from the new key range.
+        pub fn probe(self: *ScanTree, key: Key) void {
+            if (self.state == .aborted) return;
+            assert(self.state != .buffering);
+
+            // No need to move if the current range is already tighter.
+            // It can abort scanning it the key is unreachable.
+            if (key < self.key_min) {
+                if (self.direction == .descending) self.abort();
+                return;
+            } else if (self.key_max < key) {
+                if (self.direction == .ascending) self.abort();
+                return;
+            }
+
+            // Updates the scan range depending on the direction.
+            switch (self.direction) {
+                .ascending => {
+                    if (self.key_min == key) return; // No need to move
```

Comment:
Doesn't the `ZigZagMergeIterator` avoid calling `probe` on streams that already have the right key?
Could we assert that `key_min != key` here instead of returning? (Likewise for the `.descending` branch.)

Reply:
We avoid probing a scan already in the right key, but it's possible to probe the same scan twice, first as buffered and then as drained. 

This `if` does not check the current key, it only allows double-probing, so we don't have to track it during the iteration.

https://github.com/tigerbeetle/tigerbeetle/blob/7a6a44176cf9930033d690f23cd2e8dc3dd343ce/src/lsm/zig_zag_merge.zig#L162-L176